File: Aliucord/src/main/java/com/aliucord/fragments/FragmentProxy.java
/*
 * Copyright (c) 2021 Juby210 & Vendicated
 * Licensed under the Open Software License version 3.0
 */

package com.aliucord.fragments;

import android.animation.Animator;
import android.app.Activity;
import android.content.*;
import android.content.res.Configuration;
import android.os.Bundle;
import android.util.AttributeSet;
import android.view.*;
import android.view.animation.Animation;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.SharedElementCallback;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.*;
import androidx.loader.app.LoaderManager;

import com.aliucord.Main;
import com.aliucord.Utils;
import com.discord.app.AppComponent;

import java.io.FileDescriptor;
import java.io.PrintWriter;
import java.lang.reflect.Field;
import java.util.HashMap;
import java.util.Map;

import rx.subjects.Subject;

@SuppressWarnings({"CommentedOutCode", "deprecation"})
public class FragmentProxy extends Fragment implements AppComponent {
    public static final Map<String, Fragment> fragments = new HashMap<>();

    private Fragment mFragment;

//    @Override
//    public void onViewBound(View view) {
//        if (mFragment != null) mFragment.onViewBound(view);
//        super.onViewBound(view);
//    }

    @Override
    public Subject<Void, Void> getUnsubscribeSignal() {
        Fragment fragment = getmFragment();
        if (fragment instanceof AppComponent) return ((AppComponent) fragment).getUnsubscribeSignal();
        return null;
    }

    @NonNull
    @Override
    public Lifecycle getLifecycle() {
        return getmFragment().getLifecycle();
    }

    @NonNull
    @Override
    public LifecycleOwner getViewLifecycleOwner() {
        return getmFragment().getViewLifecycleOwner();
    }

    @NonNull
    @Override
    public LiveData<LifecycleOwner> getViewLifecycleOwnerLiveData() {
        return getmFragment().getViewLifecycleOwnerLiveData();
    }

    @NonNull
    @Override
    public ViewModelStore getViewModelStore() {
        return super.getViewModelStore();
    }

    @Override
    public void setArguments(@Nullable Bundle args) {
        getmFragment().setArguments(args);
    }

    @Override
    public void setInitialSavedState(@Nullable SavedState state) {
        getmFragment().setInitialSavedState(state);
    }

    @Override
    public void setTargetFragment(@Nullable Fragment fragment, int requestCode) {
        getmFragment().setTargetFragment(fragment, requestCode);
    }

    @Nullable
    @Override
    public Context getContext() {
        return getmFragment().getContext();
    }

    @Override
    public void onHiddenChanged(boolean hidden) {
        getmFragment().onHiddenChanged(hidden);
    }

    @Override
    public void setRetainInstance(boolean retain) {
        getmFragment().setRetainInstance(retain);
    }

    @Override
    public void setHasOptionsMenu(boolean hasMenu) {
        getmFragment().setHasOptionsMenu(hasMenu);
    }

    @Override
    public void setMenuVisibility(boolean menuVisible) {
        getmFragment().setMenuVisibility(menuVisible);
    }

    @Override
    public void setUserVisibleHint(boolean isVisibleToUser) {
        getmFragment().setUserVisibleHint(isVisibleToUser);
    }

    @Override
    public boolean getUserVisibleHint() {
        return getmFragment().getUserVisibleHint();
    }

    @NonNull
    @Override
    public LoaderManager getLoaderManager() {
        return getmFragment().getLoaderManager();
    }

    @Override
    public void startActivity(Intent intent) {
        getmFragment().startActivity(intent);
    }

    @Override
    public void startActivity(Intent intent, @Nullable Bundle options) {
        getmFragment().startActivity(intent, options);
    }

    @Override
    public void startActivityForResult(Intent intent, int requestCode) {
        getmFragment().startActivityForResult(intent, requestCode);
    }

    @Override
    public void startActivityForResult(Intent intent, int requestCode, @Nullable Bundle options) {
        getmFragment().startActivityForResult(intent, requestCode, options);
    }

    @Override
    public void startIntentSenderForResult(IntentSender intent, int requestCode, @Nullable Intent fillInIntent, int flagsMask, int flagsValues, int extraFlags, @Nullable Bundle options) throws IntentSender.SendIntentException {
        getmFragment().startIntentSenderForResult(intent, requestCode, fillInIntent, flagsMask, flagsValues, extraFlags, options);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        getmFragment().onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        getmFragment().onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    @Override
    public boolean shouldShowRequestPermissionRationale(@NonNull String permission) {
        return getmFragment().shouldShowRequestPermissionRationale(permission);
    }

//    @NonNull
//    @Override
//    public LayoutInflater onGetLayoutInflater(@Nullable Bundle savedInstanceState) {
////        return mFragment.onGetLayoutInflater(savedInstanceState);
//        try {
//            return super.onGetLayoutInflater(savedInstanceState);
//        } catch (Exception e) {
//            return null;
//        }
//    }
//
//    @NonNull
//    @Override
//    @SuppressLint("RestrictedApi")
//    public LayoutInflater getLayoutInflater(@Nullable Bundle savedFragmentState) {
//        return mFragment.getLayoutInflater(savedFragmentState);
//    }

    @Override
    public void onInflate(@NonNull Context context, @NonNull AttributeSet attrs, @Nullable Bundle savedInstanceState) {
        getmFragment().onInflate(context, attrs, savedInstanceState);
        super.onInflate(context, attrs, savedInstanceState);
    }

    @Override
    public void onInflate(@NonNull Activity activity, @NonNull AttributeSet attrs, @Nullable Bundle savedInstanceState) {
        getmFragment().onInflate(activity, attrs, savedInstanceState);
        super.onInflate(activity, attrs, savedInstanceState);
    }

    @Override
    public void onAttachFragment(@NonNull Fragment childFragment) {
        getmFragment().onAttachFragment(childFragment);
    }

    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        getmFragment().onAttach(context);
    }

    @Override
    public void onAttach(@NonNull Activity activity) {
        super.onAttach(activity);
        getmFragment().onAttach(activity);
    }

    @Nullable
    @Override
    public Animation onCreateAnimation(int transit, boolean enter, int nextAnim) {
        return getmFragment().onCreateAnimation(transit, enter, nextAnim);
    }

    @Nullable
    @Override
    public Animator onCreateAnimator(int transit, boolean enter, int nextAnim) {
        return getmFragment().onCreateAnimator(transit, enter, nextAnim);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        if (mFragment != null) mFragment.onCreate(savedInstanceState);
        super.onCreate(savedInstanceState);
    }

    private View mView;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        Fragment fragment = getmFragment();
        try {
            Field field = Fragment.class.getDeclaredField("mFragmentManager");
            field.setAccessible(true);
            field.set(fragment, getFragmentManager());
            Field hostField = Fragment.class.getDeclaredField("mHost");
            hostField.setAccessible(true);
            hostField.set(fragment, hostField.get(this));
        } catch (Exception e) { Main.logger.error(e); }
        mView = fragment.onCreateView(inflater, container, savedInstanceState);
        return mView;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        getmFragment().onViewCreated(view, savedInstanceState);
    }

    @Nullable
    @Override
    public View getView() {
        View v = getmFragment().getView();
        if (v == null) return mView;
        return v;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        getmFragment().onActivityCreated(savedInstanceState);
        super.onActivityCreated(savedInstanceState);
    }

    @Override
    public void onViewStateRestored(@Nullable Bundle savedInstanceState) {
        getmFragment().onViewStateRestored(savedInstanceState);
        super.onViewStateRestored(savedInstanceState);
    }

    @Override
    public void onStart() {
        getmFragment().onStart();
        super.onStart();
    }

    @Override
    public void onResume() {
        getmFragment().onResume();
        super.onResume();
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        getmFragment().onSaveInstanceState(outState);
    }

    @Override
    public void onMultiWindowModeChanged(boolean isInMultiWindowMode) {
        getmFragment().onMultiWindowModeChanged(isInMultiWindowMode);
    }

    @Override
    public void onPictureInPictureModeChanged(boolean isInPictureInPictureMode) {
        getmFragment().onPictureInPictureModeChanged(isInPictureInPictureMode);
    }

    @Override
    public void onConfigurationChanged(@NonNull Configuration newConfig) {
        getmFragment().onConfigurationChanged(newConfig);
        super.onConfigurationChanged(newConfig);
    }

    @Override
    public void onPrimaryNavigationFragmentChanged(boolean isPrimaryNavigationFragment) {
        getmFragment().onPrimaryNavigationFragmentChanged(isPrimaryNavigationFragment);
    }

    @Override
    public void onPause() {
        getmFragment().onPause();
        super.onPause();
    }

    @Override
    public void onStop() {
        getmFragment().onStop();
        super.onStop();
    }

    @Override
    public void onLowMemory() {
        getmFragment().onLowMemory();
        super.onLowMemory();
    }

    @Override
    public void onDestroyView() {
        getmFragment().onDestroyView();
        super.onDestroyView();
    }

    @Override
    public void onDestroy() {
        getmFragment().onDestroy();
        super.onDestroy();
    }

    @Override
    public void onDetach() {
        getmFragment().onDetach();
        super.onDetach();
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        getmFragment().onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public void onPrepareOptionsMenu(@NonNull Menu menu) {
        getmFragment().onPrepareOptionsMenu(menu);
    }

    @Override
    public void onDestroyOptionsMenu() {
        getmFragment().onDestroyOptionsMenu();
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        return getmFragment().onOptionsItemSelected(item);
    }

    @Override
    public void onOptionsMenuClosed(@NonNull Menu menu) {
        getmFragment().onOptionsMenuClosed(menu);
    }

    @Override
    public void onCreateContextMenu(@NonNull ContextMenu menu, @NonNull View v, @Nullable ContextMenu.ContextMenuInfo menuInfo) {
        getmFragment().onCreateContextMenu(menu, v, menuInfo);
    }

    @Override
    public void registerForContextMenu(@NonNull View view) {
        getmFragment().registerForContextMenu(view);
    }

    @Override
    public void unregisterForContextMenu(@NonNull View view) {
        getmFragment().unregisterForContextMenu(view);
    }

    @Override
    public boolean onContextItemSelected(@NonNull MenuItem item) {
        return getmFragment().onContextItemSelected(item);
    }

    @Override
    public void setEnterSharedElementCallback(@Nullable SharedElementCallback callback) {
        getmFragment().setEnterSharedElementCallback(callback);
    }

    @Override
    public void setExitSharedElementCallback(@Nullable SharedElementCallback callback) {
        getmFragment().setExitSharedElementCallback(callback);
    }

    @Override
    public void setEnterTransition(@Nullable Object transition) {
        getmFragment().setEnterTransition(transition);
    }

    @Nullable
    @Override
    public Object getEnterTransition() {
        return getmFragment().getEnterTransition();
    }

    @Override
    public void setReturnTransition(@Nullable Object transition) {
        getmFragment().setReturnTransition(transition);
    }

    @Nullable
    @Override
    public Object getReturnTransition() {
        return getmFragment().getReturnTransition();
    }

    @Override
    public void setExitTransition(@Nullable Object transition) {
        getmFragment().setExitTransition(transition);
    }

    @Nullable
    @Override
    public Object getExitTransition() {
        return getmFragment().getExitTransition();
    }

    @Override
    public void setReenterTransition(@Nullable Object transition) {
        getmFragment().setReenterTransition(transition);
    }

    @Nullable
    @Override
    public Object getReenterTransition() {
        return getmFragment().getReenterTransition();
    }

    @Override
    public void setSharedElementEnterTransition(@Nullable Object transition) {
        getmFragment().setSharedElementEnterTransition(transition);
    }

    @Nullable
    @Override
    public Object getSharedElementEnterTransition() {
        return getmFragment().getSharedElementEnterTransition();
    }

    @Override
    public void setSharedElementReturnTransition(@Nullable Object transition) {
        getmFragment().setSharedElementReturnTransition(transition);
    }

    @Nullable
    @Override
    public Object getSharedElementReturnTransition() {
        return getmFragment().getSharedElementReturnTransition();
    }

    @Override
    public void setAllowEnterTransitionOverlap(boolean allow) {
        getmFragment().setAllowEnterTransitionOverlap(allow);
    }

    @Override
    public boolean getAllowEnterTransitionOverlap() {
        return getmFragment().getAllowEnterTransitionOverlap();
    }

    @Override
    public void setAllowReturnTransitionOverlap(boolean allow) {
        getmFragment().setAllowReturnTransitionOverlap(allow);
    }

    @Override
    public boolean getAllowReturnTransitionOverlap() {
        return getmFragment().getAllowReturnTransitionOverlap();
    }

    @Override
    public void postponeEnterTransition() {
        getmFragment().postponeEnterTransition();
    }

    @Override
    public void startPostponedEnterTransition() {
        getmFragment().startPostponedEnterTransition();
    }

    @Override
    public void dump(@NonNull String prefix, @Nullable FileDescriptor fd, @NonNull PrintWriter writer, @Nullable String[] args) {
        getmFragment().dump(prefix, fd, writer, args);
    }

    private boolean didHack = false;
    public Fragment getmFragment() {
        if (mFragment == null) {
            Bundle bundle = getArguments();
            if (bundle != null) {
                String id = bundle.getString("AC_FRAGMENT_ID");
                mFragment = fragments.get(id);
                fragments.remove(id);
            }
        }

        // Horrible hack but hey it is better than crash
        if (mFragment == null) {
            if (!didHack) {
                didHack = true;
                Utils.getAppActivity().onBackPressed();
            }
            return new Fragment();
        }

        return mFragment;
    }
}

File: Aliucord/src/main/java/com/aliucord/fragments/FragmentProxy.kt
/*
 * Copyright (c) 2021 Juby210 & Vendicated
 * Licensed under the Open Software License version 3.0
 */
package com.aliucord.fragments

import android.animation.Animator
import android.app.Activity
import android.content.*
import android.content.IntentSender.SendIntentException
import android.content.res.Configuration
import android.os.Bundle
import android.util.AttributeSet
import android.view.*
import android.view.ContextMenu.ContextMenuInfo
import android.view.animation.Animation
import androidx.core.app.SharedElementCallback
import androidx.fragment.app.Fragment
import androidx.loader.app.LoaderManager
import com.aliucord.Main
import com.discord.app.AppComponent
import rx.subjects.Subject
import java.io.FileDescriptor
import java.io.PrintWriter
import java.util.*

open class FragmentProxy : Fragment(), AppComponent {
    companion object {
        val fragments: MutableMap<String, Fragment> = HashMap()
    }

    private var mFragment: Fragment? = null

    //    @Override
    //    public void onViewBound(View view) {
    //        if (mFragment != null) mFragment.onViewBound(view);
    //        super.onViewBound(view);
    //    }

    override fun getUnsubscribeSignal(): Subject<Void, Void>? {
        val fragment = getmFragment()
        return if (fragment is AppComponent) (fragment as AppComponent).unsubscribeSignal else null
    }

    override fun getLifecycle() =
        getmFragment()!!.lifecycle

    override fun getViewLifecycleOwner() =
        getmFragment()!!.viewLifecycleOwner


    override fun getViewLifecycleOwnerLiveData() =
        getmFragment()!!.viewLifecycleOwnerLiveData


    override fun setArguments(args: Bundle?) {
        getmFragment()!!.arguments = args
    }

    override fun setInitialSavedState(state: SavedState?) {
        getmFragment()!!.setInitialSavedState(state)
    }

    override fun setTargetFragment(fragment: Fragment?, requestCode: Int) {
        getmFragment()!!.setTargetFragment(fragment, requestCode)
    }

    override fun getContext() =
        getmFragment()!!.context

    override fun onHiddenChanged(hidden: Boolean) {
        getmFragment()!!.onHiddenChanged(hidden)
    }

    override fun setRetainInstance(retain: Boolean) {
        getmFragment()!!.retainInstance = retain
    }

    override fun setHasOptionsMenu(hasMenu: Boolean) {
        getmFragment()!!.setHasOptionsMenu(hasMenu)
    }

    override fun setMenuVisibility(menuVisible: Boolean) {
        getmFragment()!!.setMenuVisibility(menuVisible)
    }

    override fun setUserVisibleHint(isVisibleToUser: Boolean) {
        getmFragment()!!.userVisibleHint = isVisibleToUser
    }

    override fun getUserVisibleHint(): Boolean {
        return getmFragment()!!.userVisibleHint
    }

    override fun getLoaderManager(): LoaderManager {
        return getmFragment()!!.loaderManager
    }

    override fun startActivity(intent: Intent) {
        getmFragment()!!.startActivity(intent)
    }

    override fun startActivity(intent: Intent, options: Bundle?) {
        getmFragment()!!.startActivity(intent, options)
    }

    override fun startActivityForResult(intent: Intent, requestCode: Int) {
        getmFragment()!!.startActivityForResult(intent, requestCode)
    }

    override fun startActivityForResult(intent: Intent, requestCode: Int, options: Bundle?) {
        getmFragment()!!.startActivityForResult(intent, requestCode, options)
    }

    @Throws(SendIntentException::class)
    override fun startIntentSenderForResult(
        intent: IntentSender,
        requestCode: Int,
        fillInIntent: Intent?,
        flagsMask: Int,
        flagsValues: Int,
        extraFlags: Int,
        options: Bundle?
    ) {
        getmFragment()!!.startIntentSenderForResult(
            intent,
            requestCode,
            fillInIntent,
            flagsMask,
            flagsValues,
            extraFlags,
            options
        )
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        getmFragment()!!.onActivityResult(requestCode, resultCode, data)
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        getmFragment()!!.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    override fun shouldShowRequestPermissionRationale(permission: String) =
        getmFragment()!!.shouldShowRequestPermissionRationale(permission)

    //    @NonNull
    //    @Override
    //    public LayoutInflater onGetLayoutInflater(@Nullable Bundle savedInstanceState) {
    ////        return mFragment.onGetLayoutInflater(savedInstanceState);
    //        try {
    //            return super.onGetLayoutInflater(savedInstanceState);
    //        } catch (Exception e) {
    //            return null;
    //        }
    //    }
    //
    //    @NonNull
    //    @Override
    //    @SuppressLint("RestrictedApi")
    //    public LayoutInflater getLayoutInflater(@Nullable Bundle savedFragmentState) {
    //        return mFragment.getLayoutInflater(savedFragmentState);
    //    }
    override fun onInflate(context: Context, attrs: AttributeSet, savedInstanceState: Bundle?) {
        getmFragment()!!.onInflate(context, attrs, savedInstanceState)
        super.onInflate(context, attrs, savedInstanceState)
    }

    override fun onInflate(activity: Activity, attrs: AttributeSet, savedInstanceState: Bundle?) {
        getmFragment()!!.onInflate(activity, attrs, savedInstanceState)
        super.onInflate(activity, attrs, savedInstanceState)
    }

    override fun onAttachFragment(childFragment: Fragment) {
        getmFragment()!!.onAttachFragment(childFragment)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        getmFragment()!!.onAttach(context)
    }

    override fun onAttach(activity: Activity) {
        super.onAttach(activity)
        getmFragment()!!.onAttach(activity)
    }

    override fun onCreateAnimation(transit: Int, enter: Boolean, nextAnim: Int): Animation? {
        return getmFragment()!!.onCreateAnimation(transit, enter, nextAnim)
    }

    override fun onCreateAnimator(transit: Int, enter: Boolean, nextAnim: Int): Animator? {
        return getmFragment()!!.onCreateAnimator(transit, enter, nextAnim)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        if (mFragment != null) mFragment!!.onCreate(savedInstanceState)
        super.onCreate(savedInstanceState)
    }

    private var mView: View? = null
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val fragment = getmFragment()
        try {
            val field = Fragment::class.java.getDeclaredField("mFragmentManager").apply { isAccessible = true }
            field[fragment] = fragmentManager
            val hostField = Fragment::class.java.getDeclaredField("mHost").apply { isAccessible = true }
            hostField[fragment] = hostField[this]
        } catch (e: Exception) {
            Main.logger.error(e)
        }
        mView = fragment!!.onCreateView(inflater, container, savedInstanceState)
        return mView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        getmFragment()!!.onViewCreated(view, savedInstanceState)
    }

    override fun getView() = getmFragment()!!.view ?: mView


    override fun onActivityCreated(savedInstanceState: Bundle?) {
        getmFragment()!!.onActivityCreated(savedInstanceState)
        super.onActivityCreated(savedInstanceState)
    }

    override fun onViewStateRestored(savedInstanceState: Bundle?) {
        getmFragment()!!.onViewStateRestored(savedInstanceState)
        super.onViewStateRestored(savedInstanceState)
    }

    override fun onStart() {
        getmFragment()!!.onStart()
        super.onStart()
    }

    override fun onResume() {
        getmFragment()!!.onResume()
        super.onResume()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        getmFragment()!!.onSaveInstanceState(outState)
    }

    override fun onMultiWindowModeChanged(isInMultiWindowMode: Boolean) {
        getmFragment()!!.onMultiWindowModeChanged(isInMultiWindowMode)
    }

    override fun onPictureInPictureModeChanged(isInPictureInPictureMode: Boolean) {
        getmFragment()!!.onPictureInPictureModeChanged(isInPictureInPictureMode)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        getmFragment()!!.onConfigurationChanged(newConfig)
        super.onConfigurationChanged(newConfig)
    }

    override fun onPrimaryNavigationFragmentChanged(isPrimaryNavigationFragment: Boolean) {
        getmFragment()!!.onPrimaryNavigationFragmentChanged(isPrimaryNavigationFragment)
    }

    override fun onPause() {
        getmFragment()!!.onPause()
        super.onPause()
    }

    override fun onStop() {
        getmFragment()!!.onStop()
        super.onStop()
    }

    override fun onLowMemory() {
        getmFragment()!!.onLowMemory()
        super.onLowMemory()
    }

    override fun onDestroyView() {
        getmFragment()!!.onDestroyView()
        super.onDestroyView()
    }

    override fun onDestroy() {
        getmFragment()!!.onDestroy()
        super.onDestroy()
    }

    override fun onDetach() {
        getmFragment()!!.onDetach()
        super.onDetach()
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        getmFragment()!!.onCreateOptionsMenu(menu, inflater)
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        getmFragment()!!.onPrepareOptionsMenu(menu)
    }

    override fun onDestroyOptionsMenu() {
        getmFragment()!!.onDestroyOptionsMenu()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return getmFragment()!!.onOptionsItemSelected(item)
    }

    override fun onOptionsMenuClosed(menu: Menu) {
        getmFragment()!!.onOptionsMenuClosed(menu)
    }

    override fun onCreateContextMenu(menu: ContextMenu, v: View, menuInfo: ContextMenuInfo?) {
        getmFragment()!!.onCreateContextMenu(menu, v, menuInfo)
    }

    override fun registerForContextMenu(view: View) {
        getmFragment()!!.registerForContextMenu(view)
    }

    override fun unregisterForContextMenu(view: View) {
        getmFragment()!!.unregisterForContextMenu(view)
    }

    override fun onContextItemSelected(item: MenuItem) =
        getmFragment()!!.onContextItemSelected(item)

    override fun setEnterSharedElementCallback(callback: SharedElementCallback?) {
        getmFragment()!!.setEnterSharedElementCallback(callback)
    }

    override fun setExitSharedElementCallback(callback: SharedElementCallback?) {
        getmFragment()!!.setExitSharedElementCallback(callback)
    }

    override fun setEnterTransition(transition: Any?) {
        getmFragment()!!.enterTransition = transition
    }

    override fun getEnterTransition() =
        getmFragment()!!.enterTransition

    override fun setReturnTransition(transition: Any?) {
        getmFragment()!!.returnTransition = transition
    }

    override fun getReturnTransition() =
        getmFragment()!!.returnTransition

    override fun setExitTransition(transition: Any?) {
        getmFragment()!!.exitTransition = transition
    }

    override fun getExitTransition() =
        getmFragment()!!.exitTransition

    override fun setReenterTransition(transition: Any?) {
        getmFragment()!!.reenterTransition = transition
    }

    override fun getReenterTransition() =
        getmFragment()!!.reenterTransition

    override fun setSharedElementEnterTransition(transition: Any?) {
        getmFragment()!!.sharedElementEnterTransition = transition
    }

    override fun getSharedElementEnterTransition() =
        getmFragment()!!.sharedElementEnterTransition

    override fun setSharedElementReturnTransition(transition: Any?) {
        getmFragment()!!.sharedElementReturnTransition = transition
    }

    override fun getSharedElementReturnTransition() =
        getmFragment()!!.sharedElementReturnTransition

    override fun setAllowEnterTransitionOverlap(allow: Boolean) {
        getmFragment()!!.allowEnterTransitionOverlap = allow
    }

    override fun getAllowEnterTransitionOverlap() =
        getmFragment()!!.allowEnterTransitionOverlap

    override fun setAllowReturnTransitionOverlap(allow: Boolean) {
        getmFragment()!!.allowReturnTransitionOverlap = allow
    }

    override fun getAllowReturnTransitionOverlap() =
        getmFragment()!!.allowReturnTransitionOverlap

    override fun postponeEnterTransition() {
        getmFragment()!!.postponeEnterTransition()
    }

    override fun startPostponedEnterTransition() =
        getmFragment()!!.startPostponedEnterTransition()

    override fun dump(
        prefix: String,
        fd: FileDescriptor?,
        writer: PrintWriter,
        args: Array<String>?
    ) = getmFragment()!!.dump(prefix, fd, writer, args)

    open fun getmFragment(): Fragment? {
        if (mFragment == null) {
            if (arguments != null) {
                val id = requireArguments().getString("AC_FRAGMENT_ID")
                mFragment = fragments[id]
                fragments.remove(id)
            }
        }
        return mFragment
    }
}
